File: network_programing/c/chapter01/hello_server.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
//#include <arpa/inet.h>
//#include <sys/socket.h>
#include <winsock.h>  // #include <winsock2.h>

void error_handling(char *message);

int main(int argc, char *argv[]) {
  int serv_sock;  // 서버 소켓에 대한 파일 디스크립터 저장
  int clnt_sock;

  struct sockaddr_in serv_addr; // IP 및 Port 정보 저장
  struct sockaddr_in clnt_addr;
  // socklen_t clnt_addr_size; // POSIX 표준에 정의된 타입, 유닉스 계열이 아닌 Windows 운영체제에서는 타입이 정의되어 있지 않으므로 socklen_t 대신 int 타입을 사용해야 함
  int clnt_addr_size;

  char message[] = "Hello World!";

  if(argc != 2) {
    printf("Usage: %s <port>\n", argv[0]);
    exit(1);
  }

  serv_sock = socket(PF_INET, SOCK_STREAM, 0);  // 소켓 생성(소켓에 대한 파일 디스크립터(소켓 정보) 반환)
  if(serv_sock == -1)
    error_handling("socket() error");
  
  memset(&serv_addr, 0, sizeof(serv_addr)); // IP 및 Port 정보 저장 전 클리어(초기 상태로 세팅: 주소 초기화)
  serv_addr.sin_family = AF_INET;
  serv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
  serv_addr.sin_port = htons(atoi(argv[1]));

  if (bind(serv_sock, (struct sockaddr*) &serv_addr, sizeof(serv_addr) == -1))  // if문 실행과 동시에 bind()함수 호출: serv_sock에 저장된 소켓 정보에 해당하는 소켓에 &serv_addr를 통해 주소 정보 전달
    error_handling("bind() error");
  if (listen(serv_sock, 5) == -1) // 파일 디스크립터(소켓 정보)를 전달하면서 해당 소켓을 연결 요청
    error_handling("listen() error");

  clnt_addr_size = sizeof(clnt_addr);
  clnt_sock = accept(serv_sock, (struct sockaddr*)&clnt_addr, &clnt_addr_size); // 통신이 이루어지기 전, 서버측 OS에 요청받은 이력이 없는 지 확인 후 blocking 상태에서 머물다가 요청이 들어오면 반환을 통해 알려 줌
  if (clnt_sock == -1)
    error_handling("accept() error");
  
  write(clnt_sock, message, sizeof(message)); // 데이터 전달(Hello World! 메세지) 및 수신한 데이터를 읽음

  // 소켓 소멸
  close(clnt_sock);
  close(serv_sock);

  return 0;
}

void error_handling(char *message) {
  fputs(message, stderr);
  fputc('\n', stderr);
  exit(1);
}

File: network_programing/go/chapter01/server/hello_server.go
package main

import (
	"fmt"
	"net"
	"os"
)

func main() {
	if len(os.Args) != 2 {
		fmt.Printf("Usage: %s <port>\n", os.Args[0])
	}

	port := os.Args[1]

	// Go 언어에서는 네트워크 소켓을 생성하는 명시적인 함수 호출 및 소켓을 생성하고 설정하는 과정이 필요하지 않다
	// net.Listen() 함수는 내부적으로 소켓을 생성하고, 주어진 포트에서 연결을 수신하기 위한 리스닝 소켓을 반환한다
	// 소켓을 생성한 후 주소를 할당하는 바인딩 과정 또한 명시적으로 처리할 필요가 없다
	// net.Listen() 함수를 호출할 때, 첫 번째 인자로 사용하는 네트워크 타입과 포트 번호를 전달하면 자동으로 주소가 할당된다
	// 즉, Go 언어의 net.Listen() 함수는 주어진 네트워크 타입과 포트 번호로 소켓을 생성하고, 해당 주소에 대한 바인딩도 자동으로 수행한다
	// 굳이 직접 소켓을 직접 생성하고 싶다면, syscall 패키지를 통해 시스템 콜을 하여 수행할 수 있다
	listen, err := net.Listen("tcp", ":"+port) // 소켓 생성 후 바인딩 및 TCP 연결을 수신
	if err != nil {
		fmt.Println("Error listening:", err.Error())
		return
	}
	defer listen.Close()

	fmt.Println("Listening on port", port)

	for {
		conn, err := listen.Accept() // 연결을 허가
		if err != nil {
			fmt.Println("Error accepting connection:", err.Error())
			return
		}

		go handleConnection(conn)
	}
}

func handleConnection(conn net.Conn) {
	defer conn.Close()

	message := "Hello World!"

	_, err := conn.Write([]byte(message)) // 데이터 전달(Hello World! 메세지)
	if err != nil {
		fmt.Println("Error sending:", err.Error())
		return
	}
}
